### PR TITLE
wnbd-client: use named arguments and improve help strings

### PR DESCRIPTION
**Pull request by petrutlucian94**
At the moment, wnbd-client only uses positional arguments. This
is highly inconvenient as we're adding more arguments.

For example, the "wnbd-client map" syntax looks like this:
``wnbd-client map  <InstanceName> <HostName> <PortName> <ExportName>
                 [<SkipNBDNegotiation> <ReadOnly> <DiskSize> <BlockSize>]``

This change adds named optional arguments as well as a simple
mechanism of defining command options using Boost. This also allows
us to automatically generate better help strings.

While at it, we're adding a "--debug" option and expose soft disconnect
parameters.

### CODE REVIEW

File: wnbd-client/cmd.cpp [hunk 1]
```diff
@@ -148,15 +148,20 @@ DWORD CmdMap(
...
+    PCSTR InstanceName,
+    BOOLEAN HardRemove,
+    BOOLEAN NoHardDisonnectFallback,
+    DWORD SoftDisconnectTimeout,
+    DWORD SoftDisconnectRetryInterval)
 {
     WNBD_REMOVE_OPTIONS RemoveOptions = {0};
     RemoveOptions.Flags.HardRemove = HardRemove;
 
     // TODO: make those configurable. We should use named arguments first.
```
Comment (aserdean) [inline wnbd-client/cmd.cpp:165]: I think the comment can be dropped.

File: wnbd-client/cmd.cpp [hunk 2]
```diff
@@ -94,22 +78,35 @@ void PrintLastError()
...
         fprintf(stderr, "Missing NBD server port number.\n");
+        return ERROR_INVALID_PARAMETER;
     }
     if (SkipNegotiation && !(BlockSize && DiskSize)) {
         fprintf(stderr,
                 "The disk size and block size must be provided when "
                 "skipping NBD negotiation.\n");
+        return ERROR_INVALID_PARAMETER;
+    }
+    if (!strlen(ExportName)) {
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply wnbd-client/cmd.cpp:104]: Makes sense.

File: wnbd-client/cmd.cpp [hunk 3]
```diff
@@ -94,22 +78,35 @@ void PrintLastError()
...
+    PCSTR HostName,
     DWORD PortNumber,
-    PCHAR ExportName,
+    PCSTR ExportName,
     UINT64 DiskSize,
     UINT32 BlockSize,
     BOOLEAN SkipNegotiation,
     BOOLEAN ReadOnly)
 {
     if (!PortNumber) {
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply wnbd-client/cmd.cpp:94]: I guess we could check that it's in the [1, 65353] range.

File: wnbd-client/cmd.cpp (hunk 1 above)
Comment (petrutlucian94) [reply wnbd-client/cmd.cpp:165]: I missed this inline comment, thanks!

File: wnbd-client/cmd.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply wnbd-client/cmd.cpp]: Even better, the latest version of the PR uses std::string, simplifying this kind of checks.

File: wnbd-client/cmd.cpp [hunk 4]
```diff
@@ -21,33 +21,33 @@
...
-    std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> strconverter;
+    wstring_convert<codecvt_utf8<wchar_t>, wchar_t> strconverter;
     return strconverter.to_bytes(wstr);
 }
 
-std::string OptValToString(PWNBD_OPTION_VALUE Value)
+wstring OptValToWString(PWNBD_OPTION_VALUE Value)
 {
-    std::ostringstream stream;
+    wostringstream stream;
```
Comment (aserdean) [inline wnbd-client/cmd.cpp:38]: I'm not sure why this is being changed but if this is changed does anyone still use the function above (`to_string`)

File: wnbd-client/cmd.cpp [hunk 5]
```diff
@@ -280,72 +277,78 @@ DWORD CmdList()
...
-    DWORD Status = WnbdGetDrvOpt(Name, &Value, Persistent);
+    DWORD Status = WnbdGetDrvOpt(Name.c_str(), &Value, Persistent);
     if (Status) {
         return Status;
     }
 
-    printf("Name: %s\n", Name);
-    printf("Value: %s\n", OptValToString(&Value).c_str());
+    cout << "Name: " << Name << endl;
+    wcout << L"Value: " << OptValToWString(&Value).c_str() << endl;
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply wnbd-client/cmd.cpp:336]: Seems to work fine, shouldn't be a problem.

File: wnbd-client/cmd.cpp (hunk 4 above)
Comment (petrutlucian94) [reply wnbd-client/cmd.cpp:38]: We had to print wstrings but can just pass them to wcout. I'd keep ``to_string`` around for a bit longer, might come in handy.

File: wnbd-client/client.cpp [hunk 6]
```diff
@@ -0,0 +1,333 @@
+#include "client.h"
```
Comment (aserdean) [inline wnbd-client/client.cpp:1]: Missing license.

File: wnbd-client/client.cpp [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+#include <iostream>
+#include <iomanip>
+
+namespace po = boost::program_options;
+using namespace std;
+
+vector<Client::Command*> Client::commands;
+
+Client::Command* Client::get_command(string name) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:17]: If we want to follow the current project formatting the `{` curly bracket should be on a new line.

File: wnbd-client/client.cpp [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        }
+    }
+
+    return nullptr;
+}
+
+// Make sure to pass the EXACT SAME type that was used when defining the option,
+// otherwise Boost will throw an exception.
+template <class T>
+T safe_get_param(const po::variables_map& vm, string name, T default_val = T()) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:33]: Ditto `{`

File: wnbd-client/client.cpp [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    if (vm.count(name)) {
+        return vm[name].as<T>();
+    }
+    return default_val;
+}
+
+void Client::get_common_options(po::options_description &options) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:40]: Ditto `{`

File: wnbd-client/client.cpp [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+    options.add_options()
+        ("debug", po::bool_switch(), "Enable debug logging.");
+}
+
+void handle_common_options(po::variables_map &vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:45]: Ditto `{`

File: wnbd-client/client.cpp [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+    bool debug = safe_get_param<bool>(vm, "debug");
+
+    WnbdLogLevel log_level = debug ? WnbdLogLevelDebug : WnbdLogLevelInfo;
+    WnbdSetLogLevel(log_level);
+}
+
+DWORD Client::execute(int argc, const char** argv) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:52]: Ditto `{`

File: wnbd-client/client.cpp [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+void get_help_args(
+    po::positional_options_description &positonal_opts,
+    po::options_description &named_opts)
+{
+    positonal_opts.add("command-name", 1);
+    named_opts.add_options()
+        ("command-name", po::value<string>(), "Command name.");
+}
+
+DWORD execute_list(const po::variables_map &vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:133]: Ditto `{`

File: wnbd-client/client.cpp [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+void get_show_args(
+    po::positional_options_description &positonal_opts,
+    po::options_description &named_opts)
+{
+    positonal_opts.add("instance-name", 1);
+    named_opts.add_options()
+        ("instance-name", po::value<string>()->required(), "Disk identifier.");
+}
+
+DWORD execute_show(const po::variables_map &vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:146]: Ditto `{`

File: wnbd-client/client.cpp [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+            "This is useful for minimal NBD servers. If set, the disk size "
+            "and block size must be provided.")
+        ("disk-size", po::value<UINT64>(),
+            "The disk size. Ignored when using NBD handshake.")
+        ("block-size", po::value<UINT32>(),
+            "The block size. Ignored when using NBD handshake.")
+        ("read-only", po::bool_switch(), "Enable disk read-only mode.");
+}
+
+DWORD execute_map(const po::variables_map& vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:174]: Ditto `{`

File: wnbd-client/client.cpp [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+             "going to be disconnectd. Storage drivers can block this operation "
+             "if there are pending operations, unflushed caches or open handles. "
+             "Default: " + to_string(WNBD_DEFAULT_RM_TIMEOUT_MS / 1000) + ".").c_str())
+        ("soft-disconnect-retry-interval",
+            po::value<DWORD>()->default_value(WNBD_DEFAULT_RM_RETRY_INTERVAL_MS / 1000),
+            ("Soft disconnect retry interval in seconds. "
+             "Default: " + to_string(WNBD_DEFAULT_RM_RETRY_INTERVAL_MS / 1000) + ".").c_str());
+}
+
+DWORD execute_unmap(const po::variables_map& vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:210]: Ditto `{`

File: wnbd-client/client.cpp [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+    return CmdUnmap(
+        safe_get_param<string>(vm, "instance-name").c_str(),
+        safe_get_param<bool>(vm, "hard-disconnect"),
+        safe_get_param<bool>(vm, "no-hard-disconnect-fallback"),
+        safe_get_param<DWORD>(vm, "soft-disconnect-timeout"),
+        safe_get_param<DWORD>(vm, "soft-disconnect-retry-interval"));
+}
+
+DWORD execute_stats(const po::variables_map& vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:219]: Ditto `{`

File: wnbd-client/client.cpp [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+
+void get_list_opt_args(
+    po::positional_options_description &positonal_opts,
+    po::options_description &named_opts)
+{
+    named_opts.add_options()
+        ("persistent", po::bool_switch(), "List persistent options only.");
+}
+
+DWORD execute_list_opt(const po::variables_map& vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:241]: Ditto `{`

File: wnbd-client/client.cpp [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+    po::positional_options_description &positonal_opts,
+    po::options_description &named_opts)
+{
+    positonal_opts.add("name", 1);
+    named_opts.add_options()
+        ("name", po::value<string>()->required(), "Option name.")
+        ("persistent", po::bool_switch(), "Get the persistent value, if set.");
+}
+
+DWORD execute_get_opt(const po::variables_map& vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:256]: Ditto `{`

File: wnbd-client/client.cpp [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+{
+    positonal_opts.add("name", 1)
+                  .add("value", 1);
+    named_opts.add_options()
+        ("name", po::value<string>()->required(), "Option name.")
+        ("value", po::value<string>()->required(), "Option value.")
+        ("persistent", po::bool_switch(), "Persist the value across reboots.");
+}
+
+DWORD execute_set_opt(const po::variables_map& vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:274]: Ditto `{`

File: wnbd-client/client.cpp [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+    po::positional_options_description &positonal_opts,
+    po::options_description &named_opts)
+{
+    positonal_opts.add("name", 1);
+    named_opts.add_options()
+        ("name", po::value<string>()->required(), "Option name.")
+        ("persistent", po::bool_switch(), "Persist the value across reboots.");
+}
+
+DWORD execute_reset_opt(const po::variables_map& vm) {
```
Comment (aserdean) [inline wnbd-client/client.cpp:291]: Ditto `{`

File: wnbd-client/client.h [hunk 21]
```diff
@@ -0,0 +1,47 @@
+#pragma once
```
Comment (aserdean) [inline wnbd-client/client.h:1]: Missing license.

File: wnbd-client/usage.cc [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+        boost::trim(text);
+
+        if (!text.empty()) {
+            result += "\n";
+        }
+    }
+    return result;
+}
+
+string fmt_pos_opt_short(std::string opt_name) {
```
Comment (aserdean) [inline wnbd-client/usage.cc:56]: `{` should be on the next line.

File: wnbd-client/usage.cc [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+    ostringstream s;
+    s << "<" << opt_name << ">";
+    return s.str();
+}
+
+string fmt_opt_short(boost::shared_ptr<po::option_description> option) {
```
Comment (aserdean) [inline wnbd-client/usage.cc:62]: Ditto `{`

File: wnbd-client/usage.cc [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+    po::positional_options_description common_pos_opts;
+    po::options_description common_opts;
+    Client::get_common_options(common_opts);
+    print_option_details(common_pos_opts, common_opts, lcol_width,
+                         "Common arguments");
+
+    return 0;
+}
+
+void print_commands() {
```
Comment (aserdean) [inline wnbd-client/usage.cc:234]: Ditto `{`

File: README.md [hunk 25]
```diff
@@ -216,19 +220,22 @@ wnbd-client.exe list
...
 ```
 Pid         DiskNumber  Nbd    Owner            InstanceName
-3508        1           true   wnbd-client      test2
+3508        1           true   wnbd-client      foo
 4024        2           false  ceph-rbd-wnbd    rbd/rbd_win_10g
 ```
 
 ### Unmapping the device
 
+By default, a soft disconnect is attempted. If that fails or times out,
```
Comment (aserdean) [inline README.md:229]: Does it make sense to emphasize what a soft and hard disconnect is?

File: wnbd-client/client.cpp [hunk 26]
```diff
@@ -0,0 +1,333 @@
...
+            .options(named_opts)
+            .positional(positional_opts)
+            .run(), vm);
+        po::notify(vm);
+
+        handle_common_options(vm);
+        return command->execute(vm);
+    }
+    catch (po::required_option& e) {
+        cerr << "wnbd-client: " << e.what() << endl;
```
Comment (aserdean) [inline wnbd-client/client.cpp:102]: Does it make sense to save the binary name inside the `main` function and reuse it when needed?

File: wnbd-client/client.cpp [hunk 27]
```diff
@@ -0,0 +1,353 @@
...
+        ("instance-name", po::value<string>()->required(), "Disk identifier.")
+        ("hard-disconnect", po::bool_switch(), "Perform a hard disconnect.")
+        ("no-hard-disconnect-fallback", po::bool_switch(),
+            "Immediately return an error if the soft disconnect fails instead "
+            "of attempting a hard disconnect as fallback")
+        ("soft-disconnect-timeout",
+            po::value<DWORD>()->default_value(WNBD_DEFAULT_RM_TIMEOUT_MS / 1000),
+            ("Soft disconnect timeout in seconds. The soft disconnect operation "
+             "uses PnP to notify the Windows storage stack that the device is "
+             "going to be disconnectd. Storage drivers can block this operation "
```
Comment (aserdean) [inline wnbd-client/client.cpp:215]: s/disconnectd/disconnected/g

File: wnbd-client/client.cpp [hunk 28]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+void get_unmap_args(
+    po::positional_options_description &positonal_opts,
+    po::options_description &named_opts)
+{
+    positonal_opts.add("instance-name", 1);
+    named_opts.add_options()
+        ("instance-name", po::value<string>()->required(), "Disk identifier.")
+        ("hard-disconnect", po::bool_switch(), "Perform a hard disconnect.")
```
Comment (aserdean) [inline wnbd-client/client.cpp:207]: Should we add more information that this is a hazardous operation: i.e. Perform a hard disconnect (any pending operation will be discarded). ?

File: wnbd-client/client.cpp [hunk 29]
```diff
@@ -0,0 +1,353 @@
...
+DWORD execute_reset_opt(const po::variables_map& vm)
+{
+    return CmdResetOpt(
+        safe_get_param<string>(vm, "name").c_str(),
+        safe_get_param<bool>(vm, "persistent"));
+}
+
+Client::Command commands[] = {
+    Client::Command(
+        "version", {"-v"}, "Get the client, lib and driver version.",
```
Comment (aserdean) [inline wnbd-client/client.cpp:319]: Maybe expand lib.

File: README.md (hunk 25 above)
Comment (petrutlucian94) [reply README.md:229]: I guess adding a few words on that wouldn't hurt.

File: wnbd-client/client.cpp (hunk 26 above)
Comment (petrutlucian94) [reply wnbd-client/client.cpp:102]: Another idea would be to define a macro, something like CLIENT_NAME but I'd rather keep things simple for now.

**Review by aserdean — APPROVED** [review]

**Comment by aserdean** [comment]
Thanks for all the work!

**Pull request merged**